
<h1>Activity 2 - UK ROAD ACCIDENT DATA ANALYTICS</h1>
<hr>
<h3>Analyst: Jessie O. Mompero Jr</h3>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
from scipy.stats import f_oneway

<h2>DATABASE</h2>

In [2]:
accident_df = pd.read_csv('datasets\\uk_road_accident.csv')

In [3]:
accident_df

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h2>FILLING NULL VALUES</h2>

In [4]:
accident_df.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [5]:
accident_df['Latitude'] = accident_df['Latitude'].fillna(accident_df['Latitude'].mean())
accident_df['Longitude'] = accident_df['Longitude'].fillna(accident_df['Longitude'].mean())
accident_df['Road_Surface_Conditions'] = accident_df['Road_Surface_Conditions'].fillna(accident_df['Road_Surface_Conditions'].mode()[0])
accident_df['Road_Type'] = accident_df['Road_Type'].fillna(accident_df['Road_Type'].mode()[0])
accident_df['Urban_or_Rural_Area'] = accident_df['Urban_or_Rural_Area'].fillna(accident_df['Urban_or_Rural_Area'].mode()[0])
accident_df['Weather_Conditions'] = accident_df['Weather_Conditions'].fillna(accident_df['Weather_Conditions'].mode()[0])
accident_df.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h2>CORRECTING DATA TYPES</h2>

In [6]:
accident_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660679 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660679 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  660679 non-null  object 
 10  Road_Type                660679 non-null  object 
 11  Urban_or_Rural_Area      660679 non-null  object 
 12  Weather_Conditions       660679 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [7]:
accident_df['Index'] = accident_df['Index'].astype('category')
accident_df['Accident_Severity'] = accident_df['Accident_Severity'].astype('category')
accident_df['Accident Date'] = accident_df['Accident Date'].astype('category')
accident_df['Latitude'] = accident_df['Latitude'].astype('category')
accident_df['Light_Conditions'] = accident_df['Light_Conditions'].astype('category')
accident_df['District Area'] = accident_df['District Area'].astype('category')
accident_df['Road_Surface_Conditions'] = accident_df['Road_Surface_Conditions'].astype('category')
accident_df['Road_Type'] = accident_df['Road_Type'].astype('category')
accident_df['Urban_or_Rural_Area'] = accident_df['Urban_or_Rural_Area'].astype('category')
accident_df['Weather_Conditions'] = accident_df['Weather_Conditions'].astype('category')
accident_df['Vehicle_Type'] = accident_df['Vehicle_Type'].astype('category')
accident_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype   
---  ------                   --------------   -----   
 0   Index                    660679 non-null  category
 1   Accident_Severity        660679 non-null  category
 2   Accident Date            660679 non-null  object  
 3   Latitude                 660679 non-null  category
 4   Light_Conditions         660679 non-null  category
 5   District Area            660679 non-null  category
 6   Longitude                660679 non-null  float64 
 7   Number_of_Casualties     660679 non-null  int64   
 8   Number_of_Vehicles       660679 non-null  int64   
 9   Road_Surface_Conditions  660679 non-null  category
 10  Road_Type                660679 non-null  category
 11  Urban_or_Rural_Area      660679 non-null  category
 12  Weather_Conditions       660679 non-null  category
 13  Vehicle_Type             660679 non-null  ca

<h1>1. What district area has the most accident?</h1>

In [8]:
accident_df['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h2>Insight No 1</h2>
Insight: The Birmingham district has the most recorded accident with 13491 accident while the least is the Clackmannanshire with 91 accident.

<h1>2. Which district and what vehicle type has the highest average number of casualties per accident?</h1>

In [9]:
accident_df.groupby(['District Area', 'Vehicle_Type'])['Number_of_Casualties'].size().sort_values(ascending=False)

District Area  Vehicle_Type                
Birmingham     Car                             9600
Leeds          Car                             6875
Manchester     Car                             5248
Bradford       Car                             4749
Sheffield      Car                             4306
                                               ... 
Lincoln        Pedal cycle                        0
               Ridden horse                       0
Liverpool      Data missing or out of range       0
Teignbridge    Data missing or out of range       0
Harborough     Data missing or out of range       0
Name: Number_of_Casualties, Length: 6752, dtype: int64

<h2>Insight No 2</h2>
Insight: The Birmingham district with the car as vehicle type has the highest average number of casualties per accident.

<h1>3. What weather conditions has the most accident?</h1>

In [10]:
accident_df['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>Insight No 3</h2>
Insight: The weather conditions that recorded with the most accident is fine no high winds, which means that the weather is fine and not even high winds with 535013 records.

<h1>4. Do accidents with more vehicles tend to occur in urban or rural areas?</h1>

In [11]:
accident_df.groupby(['Urban_or_Rural_Area'])['Number_of_Vehicles'].size().sort_values(ascending=False)

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: Number_of_Vehicles, dtype: int64

<h2>Insight No 4</h2>
Insight: The Urban ares has more vehicle accident with 421678 record compared to Rural areas with only 238990 record.

<h1>5. What vehicle type has the most accident?</h1>

In [12]:
accident_df['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h2>Insight No 5</h2>
Insight: The vehicle type with the most recorded accident is the Car, with its 497992 record which is so far away from the second one Van with only 34160.

<h1>6. Which road type is most associated with fatal accidents?</h1>

In [13]:
accident_df[accident_df['Accident_Severity'] == 'Fatal']['Road_Type'].value_counts()

Road_Type
Single carriageway    6560
Dual carriageway      1815
Roundabout             142
One way street          95
Slip road               49
Name: count, dtype: int64

<h2>Insight No 6</h2>
Insight: The road type that is most associated with the fatal accidents is Single carriageway with 6560 recorded.

<h1>7. Is there a correlation between weather conditions and accident severity?</h1>

In [14]:
accident_df['Number_of_Casualties'].corr(accident_df['Number_of_Vehicles'])

np.float64(0.22888886126927563)

<h2>Insight No 7</h2>
Insight: The result is below 0.5 which suggest that there are low possibility correlation between of weather conditions and accident severity.

<h1>8. What road surface condition has the most recorded accident?</h1>

In [15]:
accident_df['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     448547
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
Name: count, dtype: int64

<h2>Insight No 8</h2>
Insight: The road surface condition that has the most recorded accident is Dry.

<h1>9. Which light condition is most common in serious accidents?</h1>

In [16]:
accident_df.groupby(['Light_Conditions', 'Accident_Severity']).size()

Light_Conditions             Accident_Severity
Darkness - lighting unknown  Fatal                    68
                             Serious                 794
                             Slight                 5622
Darkness - lights lit        Fatal                  1860
                             Serious               19130
                             Slight               108345
Darkness - lights unlit      Fatal                    45
                             Serious                 360
                             Slight                 2138
Darkness - no lighting       Fatal                  1612
                             Serious                7174
                             Slight                28651
Daylight                     Fatal                  5076
                             Serious               60759
                             Slight               419045
dtype: int64

In [17]:
accident_df[accident_df['Accident_Severity'] == 'Serious']['Light_Conditions'].mode()

0    Daylight
Name: Light_Conditions, dtype: category
Categories (5, object): ['Darkness - lighting unknown', 'Darkness - lights lit', 'Darkness - lights unlit', 'Darkness - no lighting', 'Daylight']

<h2>Insight No 9</h2>
Insight: The light codition with the most serious accidents is Daylight.

<h1>10. What road type has the most accident?</h1>

In [18]:
accident_df['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h2>Insight No 10</h2>
Insight: The road type with the most record accident is Single carriageway.

<h1>11. What weather condition does car have the most accident?</h1>

In [19]:
accident_df[accident_df['Vehicle_Type'] == 'Car']['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       403324
Raining no high winds     59940
Other                     12993
Raining + high winds       7206
Fine + high winds          6463
Snowing no high winds      4748
Fog or mist                2641
Snowing + high winds        677
Name: count, dtype: int64

<h2>Insight No 11</h2>
Insight: The weather condition that car has most accident is Fine no high winds.

<h1>12. Are fatal accidents more likely to occur during daylight or darkness?</h1>

In [20]:
accident_df.groupby(['Accident_Severity','Light_Conditions']).size()

Accident_Severity  Light_Conditions           
Fatal              Darkness - lighting unknown        68
                   Darkness - lights lit            1860
                   Darkness - lights unlit            45
                   Darkness - no lighting           1612
                   Daylight                         5076
Serious            Darkness - lighting unknown       794
                   Darkness - lights lit           19130
                   Darkness - lights unlit           360
                   Darkness - no lighting           7174
                   Daylight                        60759
Slight             Darkness - lighting unknown      5622
                   Darkness - lights lit          108345
                   Darkness - lights unlit          2138
                   Darkness - no lighting          28651
                   Daylight                       419045
dtype: int64

<h2>Insight No 12</h2>
Insight: The result shows that Fatal accidents are more likely to occur during daylight with 5076 recorded supassing all the darkness even combined.

<h1>13. What is the exact date with the most number of vehicle accident?</h1>

In [21]:
accident_df['Number_of_Vehicles'].max()

np.int64(32)

In [22]:
accident_df[accident_df['Number_of_Vehicles'] == 32]

Index Accident_Severity Accident Date   Latitude  \
497581  2009559D05333            Slight    23-12-2021  50.752096   

       Light_Conditions District Area  Longitude  Number_of_Casualties  \
497581         Daylight       Purbeck  -2.254705                     5   

        Number_of_Vehicles Road_Surface_Conditions         Road_Type  \
497581                  32            Frost or ice  Dual carriageway   

       Urban_or_Rural_Area  Weather_Conditions Vehicle_Type  
497581               Rural  Fine no high winds          Car

<h2>Insight No 13</h2>
Insight: On 23-12-2021 mark the record of most number of vehicle accident with 32 recorded accident.

<h1>14. Which vehicle type is most involved in accidents during icy road conditions?</h1>

In [23]:
accident_df[accident_df['Road_Surface_Conditions'] == 'Frost or ice']['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      14108
Van / Goods 3.5 tonnes mgw or under        972
Motorcycle over 500cc                      689
Bus or coach (17 or more pass seats)       646
Goods 7.5 tonnes mgw and over              432
Motorcycle 125cc and under                 420
Taxi/Private hire car                      384
Motorcycle 50cc and under                  224
Motorcycle over 125cc and up to 500cc      204
Other vehicle                              166
Goods over 3.5t. and under 7.5t            162
Agricultural vehicle                        59
Minibus (8 - 16 passenger seats)            40
Pedal cycle                                 10
Data missing or out of range                 1
Ridden horse                                 0
Name: count, dtype: int64

<h2>Insight No 14</h2>
Insight: During icy road conditions, the vehicle type that is most involved in accidents are Cars.

<h1>15. What accident severity has the most record?</h1>

In [24]:
accident_df['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h2>Insight No 15</h2>
Insight: The accident severity with the most record is Slight.

<h1>16. What is the most common accident severity for each weather condition?</h1>

In [25]:
accident_df.groupby(['Weather_Conditions','Accident_Severity']).size()

Weather_Conditions     Accident_Severity
Fine + high winds      Fatal                   175
                       Serious                1245
                       Slight                 7134
Fine no high winds     Fatal                  7207
                       Serious               73285
                       Slight               454521
Fog or mist            Fatal                    82
                       Serious                 483
                       Slight                 2963
Other                  Fatal                   165
                       Serious                1801
                       Slight                15184
Raining + high winds   Fatal                   145
                       Serious                1261
                       Slight                 8209
Raining no high winds  Fatal                   848
                       Serious                9468
                       Slight                69380
Snowing + high winds   Fatal             

<h2>Insight No 16</h2>
Insight: Across all weather conditions, Slight is the most common accident severity, indicating that most incidents result in minor injuries.

<h1>17. Does accident severity has correlation with the number of casualties??</h1>

In [26]:
any, p_value = f_oneway(accident_df[accident_df['Accident_Severity'] == 'Slight']['Number_of_Casualties'],accident_df[accident_df['Accident_Severity'] == 'Serious']['Number_of_Casualties'],accident_df[accident_df['Accident_Severity'] == 'Fatal']['Number_of_Casualties'])

In [27]:
p_value

np.float64(0.0)

In [28]:
pd.crosstab(accident_df['Accident_Severity'], accident_df['Number_of_Casualties'], normalize='columns')


Number_of_Casualties        1         2         3         4         5   \
Accident_Severity                                                        
Fatal                 0.009608  0.018570  0.028789  0.040516  0.058210   
Serious               0.127511  0.134249  0.171107  0.211276  0.251516   
Slight                0.862880  0.847181  0.800104  0.748207  0.690274   

Number_of_Casualties        6         7         8         9         10  ...  \
Accident_Severity                                                       ...   
Fatal                 0.080000  0.100186  0.079439  0.116071  0.120690  ...   
Serious               0.300984  0.291280  0.317757  0.294643  0.293103  ...   
Slight                0.619016  0.608534  0.602804  0.589286  0.586207  ...   

Number_of_Casualties        29   32   40   42   43   45   47   48   62   68  
Accident_Severity                                                            
Fatal                 0.000000  0.0  0.5  0.5  0.0  0.0  0.0  1.0  1.0  1.0  
Serious               0.666667  0.0  0.0  0.5  0.0  1.0  0.0  0.0  0.0  0.0  
Slight                0.333333  1.0  0.5  0.0  1.0  0.0  1.0  0.0  0.0  0.0  

[3 rows x 36 columns]

<h2>Insight No 17</h2>
Insight: The result is negative suggesting that there is no correlation between accident severity and number of casualties.

<h1>18. Which combination of road type and light condition results in the highest number of serious accidents?</h1>

In [29]:
accident_df.groupby(['Accident_Severity','Road_Type', 'Light_Conditions']).size().sort_values(ascending=False)

Accident_Severity  Road_Type           Light_Conditions           
Slight             Single carriageway  Daylight                       313828
                                       Darkness - lights lit           77808
                   Dual carriageway    Daylight                        61799
Serious            Single carriageway  Daylight                        48956
Slight             Roundabout          Daylight                        29895
                                                                       ...  
Fatal              One way street      Darkness - lighting unknown         1
                   Slip road           Darkness - lights unlit             0
                   Roundabout          Darkness - lights unlit             0
                                       Darkness - lighting unknown         0
                   One way street      Darkness - lights unlit             0
Length: 75, dtype: int64

<h2>Insight No 18</h2>
Insight: The combination of light condition of daylight and road type of roundasbout causes the highest number of serious accidents with 48956 recorded .

<h1>19. Which combination of weather and road surface conditions leads to the highest number of fatal accidents?</h1>

In [30]:
accident_df[accident_df['Accident_Severity'] == 'Fatal'].groupby(['Weather_Conditions', 'Road_Surface_Conditions']).size().sort_values(ascending=False)

Weather_Conditions     Road_Surface_Conditions
Fine no high winds     Dry                        5653
                       Wet or damp                1409
Raining no high winds  Wet or damp                 829
Raining + high winds   Wet or damp                 139
Fine no high winds     Frost or ice                125
Other                  Wet or damp                 109
Fine + high winds      Dry                         103
                       Wet or damp                  68
Fog or mist            Wet or damp                  56
Other                  Frost or ice                 43
Fog or mist            Dry                          18
Snowing no high winds  Snow                         17
Fine no high winds     Snow                         14
Other                  Dry                          12
Raining no high winds  Flood over 3cm. deep         11
Snowing no high winds  Wet or damp                   9
Fog or mist            Frost or ice                  8
Snowing no high wi

<h2>Insight No 19</h2>
Insight: The combination of weather condition of fine no high winds and road surface condition of dry causes the highest number of fatal accidents with 5653 recorded .

<h1>20. Do accidents involving more vehicles correlate or  tend to result in more casualties??</h1>

In [31]:
accident_df['Number_of_Casualties'].corr(accident_df['Number_of_Vehicles'])

np.float64(0.22888886126927563)

<h2>Insight No 20</h2>
Insight: The result is below 0.5 suggesting that there is no correlation between the number of casualties and number of vehicles.